frame/assets/src/lib.rs

```diff
@@ -269,12 +853,59 @@ impl<T: Trait> Module<T> {
 
 	/// Get the asset `id` balance of `who`.
 	pub fn balance(id: T::AssetId, who: T::AccountId) -> T::Balance {
-		<Balances<T>>::get((id, who))
+		Account::<T>::get(id, who).balance
 	}
 
 	/// Get the total supply of an asset `id`.
 	pub fn total_supply(id: T::AssetId) -> T::Balance {
-		<TotalSupply<T>>::get(id)
+		Asset::<T>::get(id).map(|x| x.supply).unwrap_or_else(Zero::zero)
+	}
+
+	/// Check the number of zombies allow yet for an asset.
+	pub fn zombie_allowance(id: T::AssetId) -> u32 {
+		Asset::<T>::get(id).map(|x| x.max_zombies - x.zombies).unwrap_or_else(Zero::zero)
+	}
+
+	fn new_account(
+		who: &T::AccountId,
+		d: &mut AssetDetails<T::Balance, T::AccountId, BalanceOf<T>>,
+	) -> Result<bool, DispatchError> {
+		let r = Ok(if frame_system::Module::<T>::account_exists(who) {
+			frame_system::Module::<T>::inc_ref(who);
+			false
+		} else {
+			ensure!(d.zombies < d.max_zombies, Error::<T>::TooManyZombies);
+			d.zombies += 1;
+			true
+		});
+		d.accounts = d.accounts.saturating_add(1);
```

Comment:
Should also be checked I think, else we risk losing track of the number of actual accounts, which could lead to deletion when users still have a balance.

Reply:
for this to overflow, it would require 4 billion individual dispatches and storage keys. there is not enough value on the chain to allow 4 billion storage keys to exist and certainly no attacker could ever amass such an amount. if they somehow did, then the merkle DB would be so incredibly slow at 1 billion keys that it would never get to 4 billion in a sensible timeframe.